keras/backend/theano_backend.py

```diff
@@ -1551,7 +1551,9 @@ def relu(x, alpha=0., max_value=None):
     return x
 
 
-def softmax(x):
+def softmax(x, axis=-1):
+    if axis != -1:
+        return T.exp(x) / T.exp(x).sum(axis=axis, keepdims=True)
```

Comment:
Could you try benchmarking this against the use of `transpose + softmax + transpose`? It may be that transposing is faster (but impossible to know in advance).

Reply:
I am a little rusty with tensor shape transforms :-). Do you mean something like `T.nnet.softmax(x.transpose(0,1,3,2)).transpose(0,1,3,2)` ?

Reply:
The exact transposition depends on the `axis` value and `x.ndim`. But we'd only need to benchmark it for one configuration. For `axis=2` and `x.ndim=4` your code looks right to me.